---
title: "Genomic_Prediction_par"
output: html_document
date: "2024-08-29"
---

This tab estimates the trait and estimated-breeding-values (EBVs) for either all individuals in a genomic dataset, or by training the model with one genomic dataset to predict the values in another. The trait and EBV information can then be used to make selections for the next breeding cycle. 

* **VCF file**: Variant Call Format (VCF) is a standard file format to store genetic variant information. The genotype (GT) data within the VCF is required for the analysis in this tab. For more details about the VCF format, see this document: https://samtools.github.io/hts-specs/VCFv4.2.pdf. 

* **Passport file**: A comma-separated values (CSV) file containing individual names (Sample_ID) in the first column and phenotype values in the subsequent columns. The phenotype column names should correspond to the phenotype ID. Example: 

<center>

|Sample_ID | Sepal.Length| Sepal.Width| Petal.Length| Petal.Width|Species |
|:---------:|:------------:|:-----------:|:------------:|:-----------:|:-------:|
|Sample_1  |          5.1|         3.5|          1.4|         0.2|versicolor   |
|Sample_2  |          4.9|         3.0|          1.4|         0.2|setosa  |
|Sample_3  |          4.7|         3.2|          1.3|         0.2|setosa  |
|Sample_4  |          4.6|         3.1|          1.5|         0.2|setosa  |
|Sample_5  |          5.0|         3.6|          1.4|         0.2|setosa  |
|Sample_6  |          5.4|         3.9|          1.7|         0.4|setosa  |

</center>

&nbsp;

* **Prediction VCF file**(optional): Variant Call Format (VCF). See above for required formatting details.

* **Species Ploidy**: Specifies the ploidy level of the species. The current analysis supports both diploids and autopolyploids. 

* **Matrix type**: Specifies the matrix type to use for the GBLUP prediction model. The choices are: 
  * **Gmatrix**: An additive relationship matrix between all samples is created using the genotype information from the VCF file 
  * **Amatrix**: An additive relationship matrix between all samples is created using a user supplied pedigree file 
  * **Hmatrix**: An additive relationship matrix between all samples is created by using the information from both the VCF file and the pedigree file 
