---
title: "PCA_par"
output: html_document
date: "2024-08-29"
---


* **VCF file**
Variant Call Format (VCF) is a standard file format to store genetic variant information. The genotype (GT) data within the VCF is required for the analysis in this tab. For more details about the VCF format, see this document: https://samtools.github.io/hts-specs/VCFv4.2.pdf.

* **Passport file** 
A comma-separated values (CSV) file containing individual names (Sample_ID) in the first column and phenotype values in the subsequent columns. The phenotype column names should correspond to the phenotype ID. Example: 

<center>

|Sample_ID | Sepal.Length| Sepal.Width| Petal.Length| Petal.Width|Species |
|:---------:|:------------:|:-----------:|:------------:|:-----------:|:-------:|
|Sample_1  |          5.1|         3.5|          1.4|         0.2|versicolor   |
|Sample_2  |          4.9|         3.0|          1.4|         0.2|setosa  |
|Sample_3  |          4.7|         3.2|          1.3|         0.2|setosa  |
|Sample_4  |          4.6|         3.1|          1.5|         0.2|setosa  |
|Sample_5  |          5.0|         3.6|          1.4|         0.2|setosa  |
|Sample_6  |          5.4|         3.9|          1.7|         0.4|setosa  |

</center>

&nbsp;

* **Species Ploidy** 
Specifies the ploidy level of the species. The current analysis supports both diploids and autopolyploids. 

* **Variable to color** 
Specifies which column to use in the passport data file to color the samples in the PCA plots by. 

* **Category to color** 
Specifies which specific item(s) within the selected column to color the samples in the PCA plots. All other samples will be a shade of grey. 

* **Color palette** 
Select which color palette to use for coloring the sample points. The different color palettes are separated by those that are color-blind friendly and those that are not. 

* **Axes to visualize** 
Choose which axes to display on the 2D PCA plot. 
